Libraries/Opc.Ua.Security.Certificates/X509Crl/X509Crl.cs

```diff
@@ -0,0 +1,345 @@
+        {
+            // Decode the Tbs and signature
+            m_signature = new X509Signature(crl);
+            // Decode the TbsCertList
+            DecodeCrl(m_signature.Tbs);
+        }
+
+        /// <summary>
+        /// Decode the Tbs of the CRL.
+        /// </summary>
+        /// <param name="tbs">The raw TbsCertList of the CRL.</param>
+        internal void DecodeCrl(byte[] tbs)
+        {
+            try
+            {
+                AsnReader crlReader = new AsnReader(tbs, AsnEncodingRules.DER);
+                var tag = Asn1Tag.Sequence;
+                var seqReader = crlReader?.ReadSequence(tag);
+                if (seqReader != null)
+                {
+                    // Version is OPTIONAL
+                    uint version = 0;
+                    var intTag = new Asn1Tag(UniversalTagNumber.Integer);
+                    var peekTag = seqReader.PeekTag();
+                    if (peekTag == intTag)
+                    {
+                        if (seqReader.TryReadUInt32(out version))
+                        {
+                            if (version != 1)
+                            {
+                                throw new AsnContentException($"The CRL contains an incorrect version {version}");
+                            }
+                        }
+                    }
+
+                    // Signature Algorithm Identifier
+                    var sigReader = seqReader.ReadSequence();
+                    var oid = sigReader.ReadObjectIdentifier();
+                    m_hashAlgorithmName = Oids.GetHashAlgorithmName(oid);
+
+                    // Issuer
+                    m_issuerName = new X500DistinguishedName(seqReader.ReadEncodedValue().ToArray());
+
+                    // thisUpdate
+                    m_thisUpdate = seqReader.ReadUtcTime().UtcDateTime;
+
+                    // nextUpdate is OPTIONAL
+                    var utcTag = new Asn1Tag(UniversalTagNumber.UtcTime);
+                    peekTag = seqReader.PeekTag();
+                    if (peekTag == utcTag)
+                    {
+                        m_nextUpdate = seqReader.ReadUtcTime().UtcDateTime;
+                    }
+
+                    var seqTag = new Asn1Tag(UniversalTagNumber.Sequence, true);
+                    peekTag = seqReader.PeekTag();
+                    if (peekTag == seqTag)
+                    {
+                        // revoked certificates
+                        var boolTag = new Asn1Tag(UniversalTagNumber.Boolean);
+                        var revReader = seqReader.ReadSequence(tag);
+                        var revokedCertificates = new List<RevokedCertificate>();
+                        while (revReader.HasData)
+                        {
+                            var crlEntry = revReader.ReadSequence();
+                            var serial = crlEntry.ReadInteger();
+                            var revokedCertificate = new RevokedCertificate(serial.ToByteArray());
```

Comment:
So RevokedCertificate is also storing serial numbers in little endian.  (Yeah, our X509Certificate classes really messed up serials, I think the hex is big-endian and the bytes is little... but it's something I always have to check against known values because it's so bad)

Reply:
Yeah exactly, it took me alot of time to get the serial number working correctly across the board with the values returned from X509Certificate2 with different endianess. Most confusing that CertificateRequest needs a big endian byte [] while GetSerialNumber() returns a little endian, SerialNumber a big endian hex etc. If you mess it up in the extensions the chain validation stops working... so I followed comparing what native extension Format() returns and matching it ... unless you tell me now storing the serial in little endian byte[] is completely off I'd keep it the way it is and be happy it works. Cheers :-)